guardian/workflow-frontend
# Replace single feature switch with multi-switchboard and add Q&A switch

## What does this change?

This PR replaces our single feature switch with a multi-switch switchboard, and adds an additional switch for the Q and A format. As before, this is available behind the shortcut <kbd>shift-f12</kbd> (<kbd>shift-fn-f12</kbd> for me on Mac).

<img width="379" alt="image" src="https://github.com/guardian/workflow-frontend/assets/34686302/cac3c9bd-34e6-40e3-8a64-c202ccaadc7e">

The switches are still stored in preferences, meaning they end up in a plain text entry in DynamoDB. However, they're now a  JSON value under the key `featureSwitches` rather than a single boolean under `featureSwitch` - along the lines of:

`{"keyTakeawaysSwitch":false,"qAndASwitch":true}`

A user won't have a value stored until they update one of the feature switches (or we add a row to DynamoDB directly).

## Note before merging:

<em>**Before merging this PR,**</em> we may want to manually re-add entries to DynamoDB for users who have opted into the old feature switch format, using the new format, to ensure continuity of access to Key Takeaways.

## How to test

1. Run workflow-frontend locally according to the instructions in the readme, or deploy to CODE. (Local uses CODE preferences so you can still access preferences there).
2. Click the page and hit <kbd>shift-f12</kbd>. Does a feature switch board appear in the top right?
3. Change one of the preferences. Does the page refresh briefly afterwards?
4. Look in the 'Create new' menu. Has the format you changed honoured your changed setting?

## Accessibility

The feature switch panel is not keyboard accessible - as it's used infrequently by a small number of users I think this is acceptable. We can manually add permissions for affected users in DynamoDB if we need to.
